app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -2,13 +2,173 @@
+        if (city == null) {
+            msgOnWrongCity();
+            return false;
+        }
+        else if (city.length() < 2 || city.length() > 30) {
+            msgOnWrongCity();
+            return false;
+        }
+        // TODO: do some other checking of input string
+        return true;
+    }
+
+    private String formatRequestWithCity(String city) {
+        String api_url = getText(R.string.weather_entry_point).toString();
+        String key = getText(R.string.weather_key).toString();
+        return api_url + "?q=" + city + "&appid=" + key + "&units=metric";
+    }
+
+    private String getTemperature(String response) {
+        String ret = null;
+
+        try {
+            JSONObject json = new JSONObject(response);
+            JSONObject main  = json.getJSONObject("main");
+            ret = main.getString("temp");
+            Log.d("myTag", "temperature = " + ret);
+        } catch (JSONException e) {
+            msgOnWrongRequest();
+            e.printStackTrace();
+        }
+
+        return ret;
+    }
+
+    private String doRequest(String request) {
+        final String[] respond = {null};
+
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                BufferedReader reader = null;
+                InputStream stream = null;
+                HttpsURLConnection connection = null;
+                try {
+                    URL weatherEndpoint = new URL(request);
+                    connection = (HttpsURLConnection) weatherEndpoint.openConnection();
+                    connection.setRequestProperty("User-Agent", getText(R.string.app_name).toString());
+                    connection.setRequestMethod("GET");
+                    connection.setReadTimeout(3000);
+                    connection.connect();
+
+                    if (connection.getResponseCode() == 200) {
+                        stream = connection.getInputStream();
+                        reader = new BufferedReader(new InputStreamReader(stream));
+                        StringBuilder buf = new StringBuilder();
+                        String line;
+                        while ((line = reader.readLine()) != null) {
+                            buf.append(line).append("\n");
+                        }
+
+                        respond[0] = buf.toString();
+
+                    } else {
+                        Log.d("myTag", "not 200");
+                    }
+
+                    if(reader != null) {
+                        reader.close();
+                    }
+
+                } catch (MalformedURLException e) {
+                    e.printStackTrace();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        });
+        thread.start();
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```

Comment:
What is the point of having separate thread if current thread will be blocked while separate thread is running?

Reply:
make doRequest() be asynchronous. It can receive callback on invocation which will have response as a parameter.